File: 0101-symmetric-tree/0101-symmetric-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function(root) {
    if (!root) return true;

    function traverse(root, left=false, path = []) {
        if (!root) {
            path.push(null);
            return '';
        };

        path.push(root.val);
        if (left) {

            traverse(root.left, left, path);
            traverse(root.right, left, path);
        } else {
            traverse(root.right, left=false, path);
            traverse(root.left, left=false, path);
            
        } 

        return path
    };

    const path1 = traverse(root.left, left=true);
    const path2 = traverse(root.right);

    if (path1.length !== path2.length) return false;
    for (let i = 0; i < path1.length; i ++) {
        if (path1[i] !== path2[i]) return false;
    };

    return true;
};

File: 0101-symmetric-tree/0101-symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        def checkSymmetric(left, right):
            if left is None and right is None:
                return True

            if left is None or right is None:
                return False
            
            if left.val != right.val:
                return False

            return checkSymmetric(left.left, right.right) and checkSymmetric(left.right, right.left)

        if not root:
            return root
        
        return checkSymmetric(root.left, root.right)